### PR TITLE
[Feature & Refinement] otaproxy, otaclient: introduce external cache dev support

### PR DESCRIPTION
**Pull request by Bodong-Yang**
## Description

> **Note**
> This PR is the first PR in a series of PRs(#244,  #245 , #246 ) that introduce extra feature to otaproxy. Each PR is implemented based on the previous PR, so these PRs should be reviewed and merged into main branch one by one.

This PR introduces the support of external cache source from external cache dev for both otaclient and otaproxy, and refines the mechanism of otaclient controlling the startup/shutdown of otaproxy.

### External cache search mechanism and implementation

When OTA update is triggered, `ota_client_stub.OTAProxyLauncher` will start subprocess launching otaproxy server in subprocess, in which during subprocess init, external cache dev with `fslabel=="ota_cache_src"' will be searched and mounted, and then otaproxy server will be configured with the using `data` dir on external cache dev as external cache source.

Please check design document for more details related to external cache dev.

### External cache source support for otaproxy

OTA proxy doesn't take care of detecting and mounting external cache source dev, instead it only takes an argument on initialization, which indicates the location of mounted external cache source.

If external cache source is available, otaproxy will first try to check if file naming with `<file_sha256>.zst`(compressed OTA file) presented, and then check if `<file_sha256>` is presented.

Please check implementation document for more details of request handling mechanism and the priority of choosing mechanism.

## Check list

- [x] local test is passed. 
- [x] VM tests are passed. 
- [x] design docs/implementation docs are prepared.

## Documents

1. [OTA cache mechanism design for otaproxy](https://tier4.atlassian.net/l/cp/yWPaqr8z)
2. [OTA proxy implementation document](https://tier4.atlassian.net/l/cp/0m2gYA7s)
3. [Performance comparison with v3.5.1](https://tier4.atlassian.net/l/cp/MErVLwBK)

## Changes

1. `app.configs`: define configs consts for external cache source support,
2. `ota_client_stub`: implement new mechanism for setting up otaproxy server in subprocess with `OTAProxyContext`, besides the previous `subprocess_init`, detecting and mounting **external cache source** is also implemented; update `OTAProxyLauncher` accordingly,
3. `ota_client_stub`: `OTAProxyLauncher` takes `OTAproxyContext` inst as injected param, and `OTAProxyContext` is initialized along with `OTAProxyLauncher` inst at `OTAClientServiceStub`. 
4. `ota_proxy.__init__`: new mechanism of configuring otaproxy at subprocess start, define `OTAProxyContextProto`, update OTACache init, update typing,
5. `ota_proxy.__main__`: CLI now takes `external-cache-folder` param,
6. `ota_proxy.cache_control`: cleanup and performance improvement,
7. `ota_proxy.ota_cache`: add support for external cache source, re-arrange request handling mechanism choosing flow; minor refinement,
8. fix up test files.

## Behavior changes

Does this PR introduce behavior change(s)?

- [x] Yes, internal behaivor (will not impact user experience).
- [x] Yes, external behaivor (will impact user experience).

### Behavior with this PR

1. otaproxy now support external cache source.
2. otaclient now has new mechanism for setting up otaproxy in subprocess.

## Breaking change

Does this PR introduce breaking change?
- [x] No.

## Related links & tickets

1. [RT4-5484](https://tier4.atlassian.net/browse/RT4-5484)

[RT4-5484]: https://tier4.atlassian.net/browse/RT4-5484?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by obi-t4 — COMMENTED** [review]
I left a comment but basically this PR is OK.

File: otaclient/ota_proxy/ota_cache.py [hunk 1]
```diff
@@ -941,17 +987,28 @@ async def retrieve_file(
...
+        url_based_id = not file_sha256_based_id
+
+        # --- case 2: if externel cache source available, try to use it --- #
+        # NOTE: if client requsts with retry_caching directive, it may indicate cache corrupted
+        #       in external cache storage, in such case we should skip the use of external cache.
+        if not cache_policy.retry_caching and (
+            _res := await self._retrieve_file_by_external_cache(
+                cache_identifier, url_based_id
+            )
+        ):
```
[the remark this answers is not in the document]
Comment (Bodong-Yang) [reply otaclient/ota_proxy/ota_cache.py]: Updated in [ee12442](https://github.com/tier4/ota-client/pull/246/commits/ee124423e7af903ab31e37c9f5cc0bb658a00b8b).

**Review by obi-t4 — APPROVED** [review]
LGTM :+1:

**Comment by Bodong-Yang** [comment]
@obi-t4 I notice some performance regression comparing to the time I did the performance comparing test. I investigate and make 2 commits to resolve this regression, please have a check, thank you!.
https://github.com/tier4/ota-client/pull/246/commits/ac0e585db06c9c7086c1cd1124351cca02476db0
https://github.com/tier4/ota-client/pull/246/commits/c580a923dd7b427acd8334af83fa5c7ad93f796a

**Comment by obi-t4** [comment]
@Bodong-Yang 

> @obi-t4 I notice some performance regression comparing to the time I did the performance comparing test. I investigate and make one commit that resolves this regression: https://github.com/tier4/ota-client/commit/ac0e585db06c9c7086c1cd1124351cca02476db0

Where is the performance regression point? It seems ac0e585db06c9c7086c1cd1124351cca02476db0 is the same as before?

**Comment by Bodong-Yang** [comment]
> @Bodong-Yang
> 
> > @obi-t4 I notice some performance regression comparing to the time I did the performance comparing test. I investigate and make one commit that resolves this regression: [ac0e585](https://github.com/tier4/ota-client/commit/ac0e585db06c9c7086c1cd1124351cca02476db0)
> 
> Where is the performance regression point? It seems [ac0e585](https://github.com/tier4/ota-client/commit/ac0e585db06c9c7086c1cd1124351cca02476db0) is the same as before? Calling url_based_hash twice is performance regression point?

It looks the same, but in VM OTA test the OTA time cost increased around 5% or more.

**Comment by obi-t4** [comment]
This time is OK but

> It looks the same

If you think so, it is noise.
If it is noise, it is better not to modify the code.
It should investigate if it is not a noise and then fix the code.

**Review by obi-t4 — APPROVED** [review]

**Comment by Bodong-Yang** [comment]
> This time is OK but
> 
> > It looks the same
> 
> If you think so, it is noise. If it is noise, it is better not to modify the code. It should investigate if it is not a noise and then fix the code.

OK, actually it is not the same. 
Take a look at L986 at fix commit, with this check, it avoids up to N calls to retrieve_file_by_external_cache method 
Also the create_cachemeta_for_request is refined so that up to N calls to url_based_hash method.

And the 5% time cost increased is definitely not noise, it is around 40 seconds increased in total OTA time cost of ~900s.

**Pull request merged**